### ISSUE TITLE
LuaRocks 3.1.3 - FreeBSD unzip returns `nil exit 1` when using the unzip -h option

### ISSUE THREAD

**Issue by RussellHaley** [luarocks/luarocks#1032]
Hi LuaRocks Team (Hisham),

Referencing prior issue here: https://github.com/luarocks/luarocks/issues/1022

I was very pleased to see the attempt to better support FreeBSD. Unfortunately the version of unzip that comes with FreeBSD does not support the -n option as used in LuaRocks. Note the error I get when I build and run luarocks from github on the v3.1.3 tag:

```
russellh@g1 ~/G/luarocks> luarocks

LuaRocks 3.1.3, the Lua package manager

<SNIP>

CONFIGURATION
	Lua version: 5.4

	Configuration files:
		System  : /usr/local/etc/luarocks/config-5.4.lua (ok)
		User    : /home/russellh/.luarocks/config-5.4.lua (not found)
		Project : /usr/home/russellh/Git/luarocks/./.luarocks/config-5.4.lua (ok)

	Rocks trees in use: 
		/usr/home/russellh/Git/luarocks/./lua_modules ("project")
		/home/russellh/.luarocks ("user")
		/usr/local ("system")

russellh@g1 ~/G/luarocks> luarocks list | less

Rocks installed for Lua 5.4
---------------------------

russellh@g1 ~/G/luarocks> luarocks install http
Installing https://luarocks.org/http-0.3-0.src.rock

Error: Failed unpacking rock file: /tmp/luarocks_luarocks-rock-http-0.3-0-zX4ddm/http-0.3-0.src.rock: 'unzip -n' program not found. Make sure unzip is installed and is available in your PATH (or you may want to edit the 'variables.UNZIP' value in file './.luarocks/config-5.4.lua')
```
Unfortunately the Lua test for `unzip -n` fails:

```

russellh@jailbird:~/freebsd/ports/devel/lua-luarocks/work/luarocks-3.1.0/src/luarocks/fs> lua53
Lua 5.3.5  Copyright (C) 1994-2018 Lua.org, PUC-Rio
> print(os.execute('unzip -n'))
Usage: unzip [-aCcfjLlnopqtuvyZ1] [-d dir] [-x pattern] zipfile
nil	exit	1

```
I've been trying to get the new version of LuaRocks pushed up to the ports collection for some time now. The 'fix' for src/luarocks/fs/unix.lua to remove the offending test can be seen in the patch I have proposed here https://reviews.freebsd.org/D17814. Other than removing the test (obviously unwise for other platforms), the best I could think of would be to test the OS flavor first, or conditionally modify that file prior to installation during the configure step? 

Incidentally, if you could add support for DragonflyBSD, and update core/cfg.lua to support the same options as the patch, I could remove all the patches once this unzip issue is resolved. That would greatly smooth the adoption of LuaRocks on FreeBSD.

My interest in fixing this is that I am currently the port maintainer for Lua 5.3 on FreeBSD and would like to adopt LuaRocks and Lua 5.4. LuaRocks is still officially on 2.x. Andrew Gierth and I are trying to push support forward for Lua on FreeBSD, but the going is slow.

Thanks!
Russ

**Comment by daurnimator** [comment]
I think you just need to provide an alternate `UNZIP` variable in your luarocks config file?

**Comment by RussellHaley** [comment]
Test unzip works...
```
russellh@fbsd-12:~> which unzip
/usr/bin/unzip
russellh@fbsd-12:~> zip tf.zip test_file*
  adding: test_file1.txt (stored 0%)
  adding: test_file2.txt (stored 0%)
russellh@fbsd-12:~> ls
ctest          git            patches        test_file1.txt tf.zip
freebsd        lua-patch      projects       test_file2.txt
russellh@fbsd-12:~> unzip -n tf.zip
Archive:  tf.zip
russellh@fbsd-12:~> rm test_file2.txt
russellh@fbsd-12:~> unzip -n tf.zip
Archive:  tf.zip
 extracting: test_file2.txt
russellh@fbsd-12:~> rm test_file2.txt
russellh@fbsd-12:~> lua53
Lua 5.3.5  Copyright (C) 1994-2018 Lua.org, PUC-Rio
> os.execute('unzip -n tf.zip')
Archive:  tf.zip
 extracting: test_file2.txt
true    exit    0
>
```

Check luarocks config file location

```
<SNIP>
        Configuration files:
                System  : /usr/local/etc/luarocks/config-5.3.lua (ok)
                User    : /home/russellh/.luarocks/config-5.3.lua (not found)

        Rocks trees in use:
                /home/russellh/.luarocks ("user")
                /usr/local ("system")
```
run luarocks

```
russellh@fbsd-12:~>
russellh@fbsd-12:~> sudo luarocks install http
Warning: Failed searching manifest: Failed extracting manifest file: 'unzip -n' program not found. Make sure unzip is installed and is available in your PATH (or you may want to edit the 'variables.UNZIP' value in file '/usr/local/etc/luarocks/config-5.3.lua')
Installing https://raw.githubusercontent.com/rocks-moonscript-org/moonrocks-mirror/master/http-0.3-0.src.rock

Error: Failed unpacking rock file: /tmp/luarocks_luarocks-rock-http-0.3-0-Wruw7V/http-0.3-0.src.rock: 'unzip -n' program not found. Make sure unzip is installed and is available in your PATH (or you may want to edit the 'variables.UNZIP' value in file '/usr/local/etc/luarocks/config-5.3.lua')
```
Modify unzip variable in config file

```
russellh@fbsd-12:~> cat /usr/local/etc/luarocks/config-5.3.lua
-- LuaRocks configuration

rocks_trees = {
   { name = "user", root = home .. "/.luarocks" };
   { name = "system", root = "/usr/local" };
}
lua_interpreter = "lua53";
variables = {
        UNZIP = "/usr/bin/unzip";
        LUA_DIR = "/usr/local";
   LUA_BINDIR = "/usr/local/bin";
}
russellh@fbsd-12:~> sudo luarocks install http
Warning: Failed searching manifest: Failed extracting manifest file: '/usr/bin/unzip' program not found. Make sure unzip is installed and is available in your PATH (or you may want to edit the 'variables.UNZIP' value in file '/usr/local/etc/luarocks/config-5.3.lua')
Installing https://raw.githubusercontent.com/rocks-moonscript-org/moonrocks-mirror/master/http-0.3-0.src.rock

Error: Failed unpacking rock file: /tmp/luarocks_luarocks-rock-http-0.3-0-HwnLEo/http-0.3-0.src.rock: '/usr/bin/unzip' program not found. Make sure unzip is installed and is available in your PATH (or you may want to edit the 'variables.UNZIP' value in file '/usr/local/etc/luarocks/config-5.3.lua')
```
Test same unzip executable path with raw lua:
```
russellh@fbsd-12:~> rm test_file2.txt
russellh@fbsd-12:~> lua
lua: Command not found.
russellh@fbsd-12:~> lua53
Lua 5.3.5  Copyright (C) 1994-2018 Lua.org, PUC-Rio
> os.execute('/usr/bin/unzip -n tf.zip')
Archive:  tf.zip
 extracting: test_file2.txt
true    exit    0
>
```

Thoughts? I'll start looking at the code when I get a chance.

**Comment by RussellHaley** [comment]
Hi, 
So if I apply the patch to src/luarocks/fs/unix/tools.lua as is in my "port diff" (https://reviews.freebsd.org/D17814), I can install fine. The is_tool_available check for unzip on line 144 is running even though the tool should be marked as installed, which was part of the 3.1.3 release (https://github.com/luarocks/luarocks/commit/fba7100c8217edec2cd41b5d8620514b274d0afd) . The error occurs when it runs unzip '-h' option on line 144. There is no -h option in FreeBSD, the command fails.

**Comment by hishamhm** [comment]
@RussellHaley could you try this and let me know if this fixes it? Thanks!

```
diff --git a/src/luarocks/fs/freebsd.lua b/src/luarocks/fs/freebsd.lua
index f72faa29..0dab7211 100644
--- a/src/luarocks/fs/freebsd.lua
+++ b/src/luarocks/fs/freebsd.lua
@@ -2,10 +2,13 @@
 local freebsd = {}
 
 local fs = require("luarocks.fs")
+local cfg = require("luarocks.core.cfg")
+
+local vars = setmetatable({}, { __index = function(_,k) return cfg.variables[k] end })
 
 function freebsd.init()
-   fs.set_tool_available("zip", true)
-   fs.set_tool_available("unzip", true)
+   fs.set_tool_available(vars.ZIP, true)
+   fs.set_tool_available(vars.UNZIP, true)
 end
 
 return freebsd
```

**Comment by RussellHaley** [comment]
Hi @hishamhm, no change:

```
russellh@fbsd-12:~/git/luarocks> ls /usr/local/share/lua/5.3/luarocks/fs/freebsd.lua
/usr/local/share/lua/5.3/luarocks/fs/freebsd.lua
russellh@fbsd-12:~/git/luarocks> cat /usr/local/share/lua/5.3/luarocks/fs/freebsd.lua | less
--- FreeBSD implementation of filesystem and platform abstractions.
local freebsd = {}

local fs = require("luarocks.fs")
local cfg = require("luarocks.core.cfg")

local vars = setmetatable({}, { __index = function(_,k) return cfg.variables[k] end })

function freebsd.init()
   fs.set_tool_available(vars.ZIP, true)
   fs.set_tool_available(vars.UNZIP, true)
end

return freebsd

russellh@fbsd-12:~/git/luarocks> sudo luarocks install http
Warning: Failed searching manifest: Failed extracting manifest file: 'unzip -n' program not found. Make sure unzip is installed and is available in your PATH (or you may want to edit the 'variables.UNZIP' value in file '/usr/local/etc/luarocks/config-5.3.lua')
Installing https://raw.githubusercontent.com/rocks-moonscript-org/moonrocks-mirror/master/http-0.3-0.src.rock

Error: Failed unpacking rock file: /tmp/luarocks_luarocks-rock-http-0.3-0-QNTen5/http-0.3-0.src.rock: 'unzip -n' program not found. Make sure unzip is installed and is available in your PATH (or you may want to edit the 'variables.UNZIP' value in file '/usr/local/etc/luarocks/config-5.3.lua')
```

**Comment by RussellHaley** [comment]
Fixed in 3.2.1. Thank you!
